libbeat/processors/annotate/kubernetes/indexing.go

```diff
@@ -0,0 +1,256 @@
+// events in a efficient way. By preindexing the metadata in the way it will be
+// checked when matching events
+type Indexer interface {
+	// GetMetadata generates event metadata for the given pod, then returns the
+	// list of indexes to create, with the metadata to put on them
+	GetMetadata(pod *corev1.Pod) []MetadataIndex
+
+	// GetIndexes return the list of indexes the given pod belongs to. This function
+	// must return the same indexes than GetMetadata
+	GetIndexes(pod *corev1.Pod) []string
+}
+
+// MetadataIndex holds a pair of index -> metadata info
+type MetadataIndex struct {
+	Index string
+	Data  common.MapStr
+}
+
+// Matcher takes a new event and returns the index
+type Matcher interface {
+	// MetadataIndex returns the index string to use in annotation lookups for the given
+	// event. A previous indexer should have generated that index for this to work
+	// This function can return "" if the event doesn't match
+	MetadataIndex(event common.MapStr) string
+}
+
+// Register contains Indexer and Matchers to use on pod indexing and event matching
+type Register struct {
+	sync.RWMutex
+	indexers []Indexer
+	matchers []Matcher
+}
+
+// NewRegister creates and returns a new Register.
+func NewRegister() *Register {
+	return &Register{
+		indexers: make([]Indexer, 0),
+		matchers: make([]Matcher, 0),
+	}
+}
+
+// AddIndexer to the register
+func (r *Register) AddIndexer(indexer Indexer) {
+	r.Lock()
+	defer r.Unlock()
+	r.indexers = append(r.indexers, indexer)
+}
+
+// AddMatcher to the register
+func (r *Register) AddMatcher(matcher Matcher) {
+	r.Lock()
+	defer r.Unlock()
+	r.matchers = append(r.matchers, matcher)
+}
+
+// GetMetadata returns the composed metadata list from all registered indexers
+func (r *Register) GetMetadata(pod *corev1.Pod) []MetadataIndex {
+	var metadata []MetadataIndex
+	r.RLock()
+	defer r.RUnlock()
+	for _, indexer := range r.indexers {
+		for _, m := range indexer.GetMetadata(pod) {
+			metadata = append(metadata, m)
+		}
+	}
+	return metadata
+}
+
+// GetIndexes returns the composed index list from all registered indexers
+func (r *Register) GetIndexes(pod *corev1.Pod) []string {
+	var indexes []string
+	r.RLock()
+	defer r.RUnlock()
+	for _, indexer := range r.indexers {
+		for _, i := range indexer.GetIndexes(pod) {
+			indexes = append(indexes, i)
+		}
+	}
+	return indexes
+}
+
+// MetadataIndex returns the index string for the first matcher from the Registry returning one
+func (r *Register) MetadataIndex(event common.MapStr) string {
+	r.RLock()
+	defer r.RUnlock()
+	for _, matcher := range r.matchers {
+		index := matcher.MetadataIndex(event)
+		if index != "" {
+			return index
+		}
+	}
+
+	// No index returned
+	return ""
+}
+
+// Implement default indexer:
+
+// GenMetadata generates default metadata for the given pod
+func GenMetadata(pod *corev1.Pod) common.MapStr {
+	labelMap := common.MapStr{}
+	for k, v := range pod.Metadata.Labels {
+		labelMap[k] = v
+	}
+	return common.MapStr{
```

Comment:
We still have the namespacing discussion open, I think we should put everything under the same `kubernetes` namespace but I was waiting to see the rest of modules to know if it makes sense